src/RevProc.cc

```diff
@@ -1905,36 +1885,55 @@ bool RevProc::ClockTick( SST::Cycle_t currentCycle ){
         DependencyClear(HartID, &(Pipeline.front().second));
       }
       Pipeline.erase(Pipeline.begin());
-      GetRegFile(HartID)->SetCost(0);
+      RegFile->SetCost(0);
     }else{
       // could not retire the instruction, bump the cost
       Pipeline.front().second.cost++;
     }
   }
-  if( GetPC() == 0x00ull ){
+  // Check for completion states and new tasks
+  // @DAVE: I feel like the below two blocks of code have some redundancy but its working and
+  //        if I comment out the first block (HartToDecode based checks) the AMO tests fail
+  //        so there must be something I'm not understanding about the two. If there isn't redundancy
+  //        great, I'll uncomment this little letter I left you
```

Comment:
Dear Ryan,

Thank you for your letter. There is some redundancy - in order to be done we need to check that there is nothing stalled waiting to decode AND that even if there is nothing to decode there is no thread with an instruction currently in flight causing us to exit early. I am curious, what if you add ` HartHasNoDependencies(HartToDecodeID)` as an additional condition in the 2nd `if` does that work?

` if( HartToExecID != _REV_INVALID_HART_ID_ && !IdleHarts[HartToExecID]
        && HartHasNoDependencies(HartToExecID)
        &&  HartHasNoDependencies(HartToDecodeID)
        && ( (nullptr == coProc) || (coProc && coProc->IsDone() )
        ) ){
      std::unique_ptr<RevThread> ActiveThread = PopThreadFromHart(HartToDecodeID);
      ActiveThread->SetState(ThreadState::DONE);
      HartsClearToExecute[HartToExecID] = false;
      HartsClearToDecode[HartToExecID] = false;
      IdleHarts[HartToExecID] = true;
      ThreadsThatChangedState.emplace(std::move(ActiveThread));
    }`

Reply:
You're a legend! Running tests now and it's looking like that works

Reply:
nvm those tests time out